internal/cmd/agent/controller/bundledeployment_controller.go

```diff
@@ -0,0 +1,241 @@
+package controller
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/go-logr/logr"
+	"github.com/rancher/fleet/internal/cmd/agent/deployer"
+	"github.com/rancher/fleet/internal/cmd/agent/deployer/cleanup"
+	"github.com/rancher/fleet/internal/cmd/agent/deployer/driftdetect"
+	"github.com/rancher/fleet/internal/cmd/agent/deployer/monitor"
+	fleetv1 "github.com/rancher/fleet/pkg/apis/fleet.cattle.io/v1alpha1"
+	"github.com/rancher/fleet/pkg/durations"
+
+	"github.com/rancher/wrangler/v2/pkg/condition"
+
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	errutil "k8s.io/apimachinery/pkg/util/errors"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/util/retry"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+)
+
+// BundleDeploymentReconciler reconciles a BundleDeployment object, by
+// deploying the bundle as a helm release.
+type BundleDeploymentReconciler struct {
+	client.Client
+	Scheme *runtime.Scheme
+
+	// LocalClient is the client for the cluster the agent is running on.
+	LocalClient client.Client
+
+	Deployer    *deployer.Deployer
+	Monitor     *monitor.Monitor
+	DriftDetect *driftdetect.DriftDetect
+	Cleanup     *cleanup.Cleanup
+
+	DefaultNamespace string
+
+	// AgentInfo is the labelSuffix used by the helm deployer
+	AgentScope string
+}
+
+var DefaultRetry = wait.Backoff{
+	Steps:    5,
+	Duration: 5 * time.Second,
+	Factor:   1.0,
+	Jitter:   0.1,
+}
+
+//+kubebuilder:rbac:groups=fleet.cattle.io,resources=bundledeployments,verbs=get;list;watch;create;update;patch;delete
+//+kubebuilder:rbac:groups=fleet.cattle.io,resources=bundledeployments/status,verbs=get;update;patch
+//+kubebuilder:rbac:groups=fleet.cattle.io,resources=bundledeployments/finalizers,verbs=update
+
+// Reconcile is part of the main kubernetes reconciliation loop which aims to
+// move the current state of the cluster closer to the desired state.
+// The Reconcile function compares the state specified by
+// the BundleDeployment object against the actual cluster state, and then
+// performs operations to make the cluster state reflect the state specified by
+// the user.
+//
+// For more details, check Reconcile and its Result here:
+// - https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.15.0/pkg/reconcile
+func (r *BundleDeploymentReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	logger := log.FromContext(ctx).WithName("bundledeployment")
+	ctx = log.IntoContext(ctx, logger)
+	key := req.String()
+
+	// get latest BundleDeployment from cluster
+	bd := &fleetv1.BundleDeployment{}
+	err := r.Get(ctx, req.NamespacedName, bd)
+	if apierrors.IsNotFound(err) {
```

Comment:
This will not work if the `BundleDeployment` is deleted when the agent is not available. In that case helm releases will be left orphaned forever.

Reply:
Fleet agent could never deal with that situation? I think it always checked all bd the next time it deployed one, as you state below.

Reply:
yes, but checking all `BundleDeployments` in each reconcilation loop, which will happens many times, is not very efficient